src/main/java/at/ac/tuwien/kr/alpha/common/program/impl/InternalProgram.java

```diff
@@ -0,0 +1,103 @@
+package at.ac.tuwien.kr.alpha.common.program.impl;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.program.AbstractProgram;
+import at.ac.tuwien.kr.alpha.common.rule.impl.InternalRule;
+import at.ac.tuwien.kr.alpha.common.rule.impl.NormalRule;
+import at.ac.tuwien.kr.alpha.grounder.FactIntervalEvaluator;
+import at.ac.tuwien.kr.alpha.grounder.Instance;
+
+/**
+ * A program in the internal representation needed for grounder and solver, i.e.: rules must have normal heads, all aggregates must be rewritten, all intervals
+ * must be preprocessed (into interval atoms), equality predicates must be rewritten
+ * 
+ * Copyright (c) 2017-2019, the Alpha Team.
+ */
+public class InternalProgram extends AbstractProgram<InternalRule> {
+
+	private final Map<Predicate, HashSet<InternalRule>> predicateDefiningRules = new LinkedHashMap<>();
+	private final Map<Predicate, LinkedHashSet<Instance>> factsByPredicate = new LinkedHashMap<>();
+	private final Map<Integer, InternalRule> rulesById = new LinkedHashMap<>();
+
+	public InternalProgram(List<InternalRule> rules, List<Atom> facts) {
+		super(rules, facts, null);
+		this.analyzeFacts(facts);
+		this.analyzeRules(rules);
+	}
+
+	protected static ImmutablePair<List<Atom>, List<InternalRule>> internalizeFactsAndRules(NormalProgram normalProgram) {
+		List<InternalRule> internalRules = new ArrayList<>();
+		List<Atom> facts = new ArrayList<>(normalProgram.getFacts());
+		for (NormalRule r : normalProgram.getRules()) {
+			if (r.getBody().isEmpty()) {
+				if (!r.getHead().isGround()) {
+					throw new IllegalArgumentException(
+							"InternalProgram does not support non-ground rules with empty bodies! (Head = " + r.getHead().toString() + ")");
+				} else {
+					facts.add(r.getHeadAtom());
+				}
+			} else {
+				internalRules.add(InternalRule.fromNormalRule(r));
+			}
+		}
+		return new ImmutablePair<>(facts, internalRules);
+	}
+
+	public static InternalProgram fromNormalProgram(NormalProgram normalProgram) {
+		ImmutablePair<List<Atom>, List<InternalRule>> factsAndRules = InternalProgram.internalizeFactsAndRules(normalProgram);
+		return new InternalProgram(factsAndRules.right, factsAndRules.left);
+	}
+
+	private void analyzeFacts(List<Atom> facts) {
+		List<Instance> tmpInstances;
+		Predicate tmpPredicate;
+		for (Atom fact : facts) { // TODO should we really unroll intervals here??
```

Comment:
Should this TODO be addressed before the pull request is accepted?

Reply:
No, I think doing this properly needs a bigger change. I removed the TODO and created issue #209 for that.